wgsl/index.bs

```diff
@@ -3060,15 +3090,82 @@ Issue:
 #### Input-output Locations TODO #### {#input-output-locations}
 TODO: *Stub*. Location-sizing of types, non-overlap among variables referenced within an entry point static call tree.
 
-### Resource interface TODO ### {#resource-interface}
+### Resource interface ### {#resource-interface}
+
+A <dfn noexport>resource</dfn> is an object,
+other than a [[#pipeline-inputs-outputs|pipeline input or output]],
+which provides access to data external to a [=shader stage=].
+Resources are shared by all invocations of the shader.
+
+There are four kinds of resources:
+
+* [=uniform buffers=]
+* [=storage buffers=]
+* textures
+* samplers
+
+The <dfn noexport>resource interface of a shader</dfn> is the set of module-scope
+resource variables [=statically accessed=] by
+[=functions in a shader stage|functions in the shader stage=].
+
+Each resource variable must be declared with both <dfn noexport>`set`</dfn>
+and <dfn noexport>`binding`</dfn> attributes.
+Together with the shader's stage, these identify the binding address
+of the resource on the shader's pipeline.
+See [[WebGPU#pipeline-layout|WebGPU &sect; GPUPipelineLayout]].
+
+Bindings must not alias within a shader stage:
+two different variables in the resource interface of a given
+shader must not have the same set and binding values, when considered as a pair of values.
+
+<table class='data'>
+  <caption>Resource variable attributes</caption>
+  <thead>
+    <tr><th>Decoraton<th>Operand<th>Description
+  </thead>
+
+  <tr><td>`set`
+      <td>non-negative i32 literal
+      <td>Bind group index
+  <tr><td>`binding`
+      <td>non-negative i32 literal
+      <td>Binding number index
+</table>
+
+### Resource layout compatibility ### {#resource-layout-compatibility}
+
+WebGPU requires that a shader's resource interface match the [[WebGPU#pipeline-layout|layout of the pipeline]]
+using the shader.
+
+Each [SHORTNAME] variable in a resource interface must be bound to a WebGPU resource with
+a compatible
+[[WebGPU#enumdef-gpubindingtype|GPUBindingType]],
+where compatibility is defined by the following table.
+<table class='data'>
+  <caption>WebGPU binding type compatibility</caption>
+  <thead>
+    <tr><th>[SHORTNAME] resource
+        <th>WebGPU [[WebGPU#enumdef-gpubindingtype|GPUBindingType]]
+  </thead>
+  <tr><td>[=uniform buffer=]
+      <td>[[WebGPU#dom-gpubindingtype-uniform-buffer|uniform-buffer]]
+  <tr><td>[=storage buffer=] with `access(read_write)` attribute
+      <td>[[WebGPU#dom-gpubindingtype-storage-buffer|storage-buffer]]
+  <tr><td>[=storage buffer=] with `access(read)` attribute
```

Comment:
The `access(read)` is not in the spec at this point. We, technically, only have read write default storage buffers.